File: python/two-pointers/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        left, right = 0, len(height) - 1
        maxleft, maxright = height[left], height[right]
        answer = 0
        while left < right:
            answer += min(maxleft, maxright)
            if height[left] < height[right]:
                answer -= height[left]
                left += 1
            else:
                answer -= height[right]
                right -= 1

            maxleft = max(maxleft, height[left])
            maxright = max(maxright, height[right])

        return answer

File: c++/two-pointers/trapping-rain-water.cc
class Solution {
public:
    int trap(vector<int>& height) {
        int left = 0;
        int right = height.size() - 1;

        int max_left = height[left];
        int max_right = height[right];

        int water = 0;
        while (left < right) {
            water += std::min(max_left, max_right);
            if (height[left] > height[right]) {
                water -= height[right];
                right -= 1;
                max_right = std::max(max_right, height[right]);
            } else {
                water -= height[left];
                left += 1;
                max_left = std::max(max_left, height[left]);
            }
        }

        return water;
    }
};
